lib/semantics/resolve-names.cc

```diff
@@ -2510,27 +2529,131 @@ void DeclarationVisitor::Post(const parser::LengthSelector &x) {
   }
 }
 
-void DeclarationVisitor::Post(const parser::DeclarationTypeSpec::Class &x) {
-  // created by default with TypeDerived; change to ClassDerived
-  GetDeclTypeSpec()->set_category(DeclTypeSpec::ClassDerived);
+bool DeclarationVisitor::Pre(const parser::DeclarationTypeSpec::Type &x) {
+  SetDeclTypeSpecCategory(DeclTypeSpec::Category::TypeDerived);
```

Comment:
This probably isn't necessary as the category defaults to TypeDerived.

Reply:
Will it always be so, do you think?

Reply:
I think so, but can't be sure. My current pull request has the same change: add `state_.category`, initialize it to `TypeDerived`, change it on seeing `parser::DeclarationTypeSpec::Class`. But I didn't also have this `Pre` function.

An alternative would be to make the category optional, set it on both `parser::DeclarationTypeSpec::Type` and `parser::DeclarationTypeSpec::Class`, and check that it is set when it is needed. That might be a little more robust.

Reply:
I'll change the Set of the category to a CHECK(Get).

Reply:
Sounds good